src/wagtail_vector_index/ai_utils/backends/base.py

```diff
@@ -0,0 +1,222 @@
+class BaseChatConfig(BaseConfig[ChatConfigSettings]):
+    pass
+
+
+@dataclass(kw_only=True)
+class BaseEmbeddingConfig(BaseConfig[EmbeddingConfigSettings]):
+    embedding_output_dimensions: int
+
+    @classmethod
+    def from_settings(
+        cls,
+        config: EmbeddingConfigSettings,
+        *,
+        text_splitter_class: type[TextSplitterProtocol],
+        text_splitter_length_calculator_class: type[
+            TextSplitterLengthCalculatorProtocol
+        ],
+        **kwargs: Any,
+    ) -> Self:
+        embedding_output_dimensions = cls.get_embedding_output_dimensions(
+            model_id=config["MODEL_ID"],
+            custom_value=config.get("EMBEDDING_OUTPUT_DIMENSIONS"),
+        )
+        kwargs.setdefault("embedding_output_dimensions", embedding_output_dimensions)
+        return super().from_settings(
+            config=config,
+            text_splitter_class=text_splitter_class,
+            text_splitter_length_calculator_class=text_splitter_length_calculator_class,
+            **kwargs,
+        )
+
+    @classmethod
+    def get_embedding_output_dimensions(
+        cls, *, model_id: str, custom_value: int
+    ) -> int:
+        if custom_value is not None:
+            try:
+                return int(custom_value)
+            except ValueError as e:
+                raise ImproperlyConfigured(
+                    f'"EMBEDDING_OUTPUT_DIMENSIONS" is not an "int", it is a "{type(custom_value)}".'
+                ) from e
+        try:
+            return embeddings.get_default_embedding_output_dimensions(model_id=model_id)
+        except embeddings.EmbeddingOutputDimensionsNotFound as e:
+            raise ImproperlyConfigured(
+                f'"EMBEDDING_OUTPUT_DIMENSIONS" is not configured for model "{model_id}".'
+            ) from e
+
+
+AnyBackendConfig = TypeVar("AnyBackendConfig", bound=BaseConfig)
+ChatBackendConfig = TypeVar("ChatBackendConfig", bound=BaseChatConfig)
+EmbeddingBackendConfig = TypeVar("EmbeddingBackendConfig", bound=BaseEmbeddingConfig)
+
+
+class BaseBackend(Generic[AnyBackendConfig]):
+    config_cls: ClassVar[type[BaseConfig]]
+    config: AnyBackendConfig
+
+    def __init__(self, *, config: AnyBackendConfig) -> None:
+        self.config = config
+
+    def get_text_splitter(self) -> TextSplitterProtocol:
+        return self.config.text_splitter_class(
+            chunk_size=self.config.token_limit,
+            chunk_overlap=self.config.chunk_overlap_characters,
+            length_function=self.get_splitter_length_calculator().get_splitter_length,
+        )
+
+    def get_splitter_length_calculator(self) -> TextSplitterLengthCalculatorProtocol:
+        return self.config.text_splitter_length_calculator_class()
+
+
+class BaseChatBackend(BaseBackend[ChatBackendConfig]):
+    config_cls: ClassVar[type[BaseChatConfig]]
+    config: ChatBackendConfig
+
+    def chat(self, *, user_messages: Sequence[str]) -> AIResponse:
+        ...
+
+    def prompt_with_context(
```

Comment:
**Issue:** Remove this. We don't need prompt_with_context because that's only used by Wagtail AI.